lexis/render/ColorMap.h

```diff
@@ -0,0 +1,249 @@
+     * @returns colors that are interpolated for given control points range.
+     */
+    template< typename T >
+    LEXIS_API void sampleColors( T* data, size_t count, float min, float max, T emptyColor );
+
+    /**
+     * Adds a control point to a given channel. If there is already a control point for the
+     * given cp.x, it is replaced with the given control point.
+     * @param cp control points
+     */
+    LEXIS_API void addControlPoint( const detail::ControlPoint& cp, Channel channel );
+
+    /** returns true if there are no control points in any channel */
+    LEXIS_API bool isEmpty() const;
+
+private:
+
+    ::zerobuf::Vector< detail::ControlPoint >& _getControlPoints( Channel channel );
+    void _sortChannels();
+    bool _isSorted;
+};
+
+template< typename T >
+Colors< T > ColorMap::sampleColors( size_t count )
+{
+    return sampleColors< T >( count,
+                              -std::numeric_limits< float >::infinity(),
+                              std::numeric_limits< float >::infinity(),
+                              static_cast< T >( 0 ));
+}
+
+template< typename T >
+void ColorMap::sampleColors( T* data, size_t count )
+{
+    return sampleColors< T >( data,
+                              count,
+                              -std::numeric_limits< float >::infinity(),
+                              std::numeric_limits< float >::infinity(),
+                              static_cast< T >( 0 ));
+}
+
+template< typename T >
+Colors< T > ColorMap::sampleColors( size_t count, float min, float max, T emptyColor )
+{
+    Colors< T > colors( count, { emptyColor, emptyColor, emptyColor, emptyColor });
+    sampleColors( reinterpret_cast< T* >( colors.data( )), count, min, max, emptyColor );
+    return colors;
+}
+
+template< typename T >
+void ColorMap::sampleColors( T* data, size_t count, float min, float max, T emptyColor )
+{
+    _sortChannels();
+    const Channel channels[] = { Channel::red, Channel::green, Channel::blue, Channel::alpha };
+    for( const auto& channel: channels )
+    {
+        ::zerobuf::Vector< detail::ControlPoint >& cps = _getControlPoints( channel );
+        const size_t cpsSize = cps.size();
+        if( cps.empty( )) // No element, do nothing
+            continue;
+        else if( cpsSize == 1 ) // 1 element
+        {
+            for( size_t sample = 0; sample < count; ++sample )
+                data[ sample * ( size_t )Channel::nChannels + ( size_t )channel ] = cps[ 0 ].getY();
+            continue;
+        }
+
+        const detail::ControlPoint& begin = cps[ 0 ];
+        const detail::ControlPoint& end = cps[ cpsSize - 1 ];
+
+        const float cpMinX = begin.getX();
+        const float cpMaxX = end.getX();
+        const float minX = min == -std::numeric_limits< float >::infinity() ? cpMinX : min;
+        const float maxX = max == std::numeric_limits< float >::infinity() ? cpMaxX : max;
```

Comment:
Without fully understanding the whole code this seems wrong. If I understand this correctly, when the range is not user given (-inf, inf case), you will pick a different pair of minX maxX values per channel, which is completely wrong. You have to pick the minimum and maximum from all the channels. You use the emptyColor (which should be renamed to something else, because the same value is used for all channels) to fill the missing points when you're outside the range of a given channel.